analyzers/src/SonarAnalyzer.Common/SyntaxTrackers/SyntaxTrackerBase.cs

```diff
@@ -24,29 +24,42 @@
 
 namespace SonarAnalyzer.Helpers
 {
-    public abstract class SyntaxTrackerBase<TSyntaxKind>
+    public abstract class SyntaxTrackerBase<TSyntaxKind, TCondition> : TrackerBase
         where TSyntaxKind : struct
     {
-        private readonly IAnalyzerConfiguration analyzerConfiguration;
+        protected abstract TSyntaxKind[] TrackedSyntaxKinds { get; }
 
-        protected DiagnosticDescriptor Rule { get; }
+        protected abstract BaseContext IsTracked(SyntaxNode expression, SemanticModel semanticModel, TCondition[] conditions, out Location location);
 
-        protected abstract GeneratedCodeRecognizer GeneratedCodeRecognizer { get; }
+        protected SyntaxTrackerBase(IAnalyzerConfiguration analyzerConfiguration, DiagnosticDescriptor rule) : base(analyzerConfiguration, rule) { }
 
-        protected abstract TSyntaxKind[] TrackedSyntaxKinds { get; }
+        public void Track(SonarAnalysisContext context, params TCondition[] conditions) =>
+            Track(context, new object[0], conditions);
 
-        protected SyntaxTrackerBase(IAnalyzerConfiguration analyzerConfiguration,
-            DiagnosticDescriptor rule)
+        public void Track(SonarAnalysisContext context, object[] diagnosticMessageArgs, params TCondition[] conditions)
         {
-            this.analyzerConfiguration = analyzerConfiguration;
-            this.Rule = rule;
-        }
+            context.RegisterCompilationStartAction(
+              c =>
+              {
+                  if (IsEnabled(c.Options))
+                  {
+                      c.RegisterSyntaxNodeActionInNonGenerated(
+                          GeneratedCodeRecognizer,
+                          TrackAndReportIfNecessary,
+                          TrackedSyntaxKinds);
+                  }
+              });
 
-        protected bool IsEnabled(AnalyzerOptions options)
-        {
-            analyzerConfiguration.Initialize(options);
+            void TrackAndReportIfNecessary(SyntaxNodeAnalysisContext c)
+            {
+                var baseContext = IsTracked(c.Node, c.SemanticModel, conditions, out var location);
 
-            return analyzerConfiguration.IsEnabled(Rule.Id);
+                if (baseContext != null)
```

Comment:
After addressing other comments, this could look something like this
```suggestion
                if (CreateContext(c.Node, c.SemanticModel, conditions) is {} context
                    && context.PrimaryLocation != null)
```